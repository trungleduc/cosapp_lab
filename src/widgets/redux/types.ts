/**
 * Interface of redux store
 */
export interface StateInterface {
  /** Global configuration */
  systemConfig: { [key: string]: any };
  /** State of activated panel */
  mainState: PanelStatusInterface;
  saveSignal: number;
  /** State of system architecture panel*/
  systemArch: {
    lockStatus: boolean; // lock status of diagram
    /** State of system tree structure */
    systemTree: {
      nodeData: any; // Data of each node
      nodePath: string[]; // Path from root to each node
      selectedNode: string; // Current selected node for editing
    };
    /** State which holds the data of each system */
    systemData: SystemDataInterface;
    /** State which holds the data of for diagram */
    systemGraph: SystemGraphInterface;
    /** State of system pbs view */
    systemPBS: { [key: string]: { visible: boolean; position: Array<number> } };
    /**Signal for node filtering */
    systemPBSUpdated: number;
  };
  /**State of dashboard panel */
  dashboardState: {
    computing: boolean; // True if a driver is running, False otherwise.
    timeStepList: string[]; // List of time step available.
    data3D: { [key: string]: any[] }; // Not used yet.
    variableData: {[key: string]: any;}; // System variable data received from backend.
    portMetaData: { [key: string]: any }; // Port meta data received from backend.
    selectedVariable: { [key: string]: number }; // Controller variable and its value.
    computedResult: { [key: string]: any }; // Value of all variable, used for plot
    recorderData: { [key: string]: any }; // Cosapp recorder
    driverData: { [key: string]: any }; // Cosapp driver data
  };
}

export interface IDict<T> {[key:string]: T}
export interface SystemDataInterface {
  mainData: {
    // key is name of system
    [key: string]: {
      user: UserListType[];
      class: string;
      portList: string[];
    };
  };
  portData: {
    //key in form of "systemName/portName"
    [key: string]: {
      variable: string[];
      connection: string[]; // Destination port in form of "systemName/portName"
    };
  };
}

export interface PanelStatusInterface {
  [key: string]: [boolean, string];
}

export interface SystemGraphInterface {
  systemGraphData: {
    [key: string]: {
      inPort: Array<string>;
      outPort: Array<string>;
      connections: Array<Array<any>>;
      position?: Array<number>;
    };
  };
  systemList: Array<string>;
  graphJsonData: { [key: string]: any };
  updateData: { [key: string]: any };
}

export enum Action {
  SAVE_STATE = 'SAVE_STATE',
  RESET_STORE = 'RESET_STORE',
  SWITCH_PANEL = 'SWITCH_PANEL',
  SELECT_NODE = 'SELECT_NODE',
  ADD_NODE = 'ADD_NODE',
  REMOVE_NODE = 'REMOVE_NODE',
  UPDATE_TREE = 'UPDATE_TREE',
  ARCH_SAVE_GRAPH_JSON = 'ARCH_SAVE_GRAPH_JSON',
  ARCH_SAVE_GRAPH_POSITION = 'ARCH_SAVE_GRAPH_POSITION',
  ARCH_UPDATE_CONNECTION_GRAPH = 'ARCH_UPDATE_CONNECTION_GRAPH',
  ARCH_FILTER_NODE = 'ARCH_FILTER_NODE',
  LOCK_GRAPH = 'LOCK_GRAPH',
  DASHBOARD_ADD_3D_DATA = 'DASHBOARD_ADD_3D_DATA',
  GET_SERVER_DATA = 'GET_SERVER_DATA',
  DASHBOARD_TOGGLE_RUN = 'DASHBOARD_TOGGLE_RUN',
  DASHBOARD_ADD_CONTROLLER = 'DASHBOARD_ADD_CONTROLLER',
  DASHBOARD_SET_CONTROLLER = 'DASHBOARD_SET_CONTROLLER',
  DASHBOARD_REMOVE_CONTROLLER = 'DASHBOARD_REMOVE_CONTROLLER',
  DASHBOARD_UPDATE_CONTROLLER = 'DASHBOARD_UPDATE_CONTROLLER',
  DASHBOARD_UPDATE_COMPUTED_RESULT = 'DASHBOARD_UPDATE_COMPUTED_RESULT',
  DASHBOARD_UPDATE_RECORDER_DATA = 'DASHBOARD_UPDATE_RECORDER_DATA',
  DASHBOARD_UPDATE_DRIVER_DATA = 'DASHBOARD_UPDATE_DRIVER_DATA'
}

export interface SaveState {
  type: Action.SAVE_STATE;
}
export interface ResetStore {
  type: Action.RESET_STORE;
}
export interface MainSwitchPanel {
  type: Action.SWITCH_PANEL;
  panelName: string;
}

export interface UserListType {
  uid: number;
  name: string;
}

export interface ServerGetData {
  type: Action.GET_SERVER_DATA;
  data: any;
}

export interface ArchSaveGraphJson {
  type: Action.ARCH_SAVE_GRAPH_JSON;
  data: { [key: string]: any };
}

export interface ArchSaveGraphPosition {
  type: Action.ARCH_SAVE_GRAPH_POSITION;
  data: { [key: string]: { visible: boolean; position: Array<number> } };
}

export interface ArchFilterNode {
  type: Action.ARCH_FILTER_NODE;
  data: Array<string>;
  selected: Array<string>;
}

export interface ArchUpdateConnectionGraph {
  type: Action.ARCH_UPDATE_CONNECTION_GRAPH;
  data: { [key: string]: string };
  root: string;
}

export interface ArchLockGraph {
  type: Action.LOCK_GRAPH;
  status: boolean;
}

export interface ArchSelectNode {
  type: Action.SELECT_NODE;
  nodeId: string;
}

export interface ArchAddNode {
  type: Action.ADD_NODE;
  nodeName: string;
  nodePath: any;
  nodeData: {
    personName: UserListType[];
    sysClass: string;
    portList: {
      portName: string;
      portDirection: number;
      portClass: string;
    }[];
  };
}

export interface ArchRemoveNode {
  type: Action.REMOVE_NODE;
  nodeData: any;
  nodePath: any;
}

export interface ArchUpdateTree {
  type: Action.UPDATE_TREE;
  treeData: any;
  updatePbs: boolean;
}
export interface DashboardAddData {
  type: Action.DASHBOARD_ADD_3D_DATA;
  data: any;
}

export interface DashboardToggleRun {
  type: Action.DASHBOARD_TOGGLE_RUN;
}

export interface DashboardSetController {
  type: Action.DASHBOARD_SET_CONTROLLER;
  data: { [key: string]: number };
}
export interface DashboardAddController {
  type: Action.DASHBOARD_ADD_CONTROLLER;
  variableName: string;
  value: number;
}

export interface DashboardUpdateController {
  type: Action.DASHBOARD_UPDATE_CONTROLLER;
  variableName: string;
  value: number;
}

export interface DashboardRemoveController {
  type: Action.DASHBOARD_REMOVE_CONTROLLER;
  variableName: string;
}

export interface DashboardUpdateComputedResult {
  type: Action.DASHBOARD_UPDATE_COMPUTED_RESULT;
  data: {
    variable: { [key: string]: any };
    recorder: { [key: string]: any };
    driver: { [key: string]: any };
  };
}

export type ActionType =
  | ResetStore
  | SaveState
  | MainSwitchPanel
  | ServerGetData
  | ArchSaveGraphJson
  | ArchSaveGraphPosition
  | ArchFilterNode
  | ArchLockGraph
  | ArchUpdateTree
  | ArchSelectNode
  | ArchAddNode
  | ArchRemoveNode
  | ArchUpdateConnectionGraph
  | DashboardAddData
  | DashboardToggleRun
  | DashboardSetController
  | DashboardAddController
  | DashboardRemoveController
  | DashboardUpdateController
  | DashboardUpdateComputedResult;
