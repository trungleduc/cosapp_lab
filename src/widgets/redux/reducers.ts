import { Action, ActionType, StateInterface } from './types';
import * as ActionFunc from './actions';

export const initialState: StateInterface = {
  systemConfig: {},
  mainState: {
    panel_0: [false, 'Project_Overview'],
    panel_1: [true, 'System_Arc'],
    panel_2: [false, 'Module_Creator'],
    panel_3: [false, 'Dashboard'],
    panel_4: [false, 'User_Management'],
    panel_5: [false, 'Module_Library'],
  },
  saveSignal: 0,
  systemArch: {
    lockStatus: false,
    systemTree: {
      nodeData: [{ title: 'Root' }],
      nodePath: [],
      selectedNode: null,
    },
    systemData: { mainData: {}, portData: {} },
    systemGraph: {
      systemGraphData: {
        Root: {
          inPort: ['inwards'],
          outPort: ['outwards'],
          connections: [],
        },
      },
      systemList: ['Root'],
      graphJsonData: {},
      updateData: {},
    },
    systemPBS: {},
    systemPBSUpdated: 0,
  },
  dashboardState: {
    computing: false,
    timeStepList: ['None'],
    data3D: {},
    variableData: {},
    portMetaData: {},
    selectedVariable: {},
    computedResult: {},
    recorderData: {},
    driverData : {}
  },
};

export function rootReducer(
  state: StateInterface = initialState,
  action: ActionType
): StateInterface {
  switch (action.type) {
    case Action.RESET_STORE: {
      return initialState;
    }

    case Action.SAVE_STATE: {
      return ActionFunc.saveState_(state);
    }

    case Action.SWITCH_PANEL: {
      return ActionFunc.mainSwitchPanel_(state, action);
    }

    case Action.LOCK_GRAPH: {
      return {
        ...state,
        systemArch: {
          ...state.systemArch,
          lockStatus: action.status,
        },
      };
    }

    case Action.GET_SERVER_DATA: {
      return ActionFunc.serverGetData_(state, action);
    }

    case Action.ARCH_SAVE_GRAPH_JSON: {
      return ActionFunc.archSaveGraphJson_(state, action);
    }

    case Action.ARCH_SAVE_GRAPH_POSITION: {
      return ActionFunc.archSaveGraphPosition_(state, action);
    }

    case Action.ARCH_UPDATE_CONNECTION_GRAPH: {
      return ActionFunc.archUpdateConnectionGraph_(state, action);
    }

    case Action.ARCH_FILTER_NODE: {
      return ActionFunc.archFilterNode_(state, action);
    }

    case Action.SELECT_NODE: {
      return ActionFunc.archSelectNode_(state, action);
    }

    case Action.ADD_NODE: {
      return ActionFunc.archAddNode_(state, action);
    }

    case Action.UPDATE_TREE: {
      return ActionFunc.archUpdateTreeData_(state, action);
    }

    case Action.REMOVE_NODE: {
      return ActionFunc.archRemoveNode_(state, action);
    }

    case Action.DASHBOARD_ADD_3D_DATA: {
      return ActionFunc.dashboardAdd3DData_(state, action);
    }

    case Action.DASHBOARD_TOGGLE_RUN: {
      return {
        ...state,
        dashboardState: {
          ...state.dashboardState,
          computing: !state.dashboardState.computing,
        },
      };
    }

    case Action.DASHBOARD_SET_CONTROLLER: {
      return ActionFunc.dashboardSetController_(state, action);
    }
    case Action.DASHBOARD_ADD_CONTROLLER: {
      return ActionFunc.dashboardAddController_(state, action);
    }

    case Action.DASHBOARD_REMOVE_CONTROLLER: {
      return ActionFunc.dashboardRemoveController_(state, action);
    }

    case Action.DASHBOARD_UPDATE_CONTROLLER: {
      return ActionFunc.dashboardUpdateController_(state, action);
    }

    case Action.DASHBOARD_UPDATE_COMPUTED_RESULT: {
      return ActionFunc.dashboardUpdateComputedResult_(state, action);
    }

  }

  return state;
}
