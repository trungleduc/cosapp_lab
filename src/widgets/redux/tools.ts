import { getNodeAtPath, walk } from 'react-sortable-tree';

/**
 *
 * Helper function to call `fn` each `ms` millisecond
 * @export
 * @param {*} fn
 * @param {*} ms
 * @param {*} number
 * @returns {*}
 */

export function debounce(fn: any, ms: number): any {
  let timer;
  return () => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      fn();
    }, ms);
  };
}

/**
 *
 *
 * @export
 * @param {*} treeData
 * @param {number[]} path
 * @returns
 */
export function getNodeFullPath(treeData: any, path: number[]) {
  const getNodeKey = (obj: { treeIndex: any }) => obj.treeIndex;
  let fullPath: string;
  for (let index = 0; index < path.length; index++) {
    const newPath = path.slice(0, index + 1);
    const currentNode = getNodeAtPath({
      treeData: treeData,
      path: newPath,
      getNodeKey: getNodeKey
    });
    const currentTitle = currentNode.node.title;
    if (index == 0) {
      fullPath = currentTitle as string;
    } else {
      fullPath = fullPath.concat('.' + currentTitle);
    }
  }
  return fullPath;
}

export function updateNodeId(treeData: any, oldId: string, newId: string) {
  const getNodeKey = (obj: { treeIndex: any }) => obj.treeIndex;
  const ret = {};
  const callback = item => {
    const nodeId: string = item.node.id;
    if (nodeId.startsWith(oldId)) {
      const newNodeId = nodeId.replace(oldId, newId);
      item.node.id = newNodeId;
      ret[nodeId] = newNodeId;
    }
  };
  walk({ treeData, getNodeKey, callback });

  return ret;
}
