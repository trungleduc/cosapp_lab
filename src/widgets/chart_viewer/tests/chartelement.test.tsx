//@ts-ignore
global.URL.createObjectURL = jest.fn();
HTMLCanvasElement.prototype.getContext = jest.fn();
import { Provider } from 'react-redux';
import { initialState } from '../../redux/reducers';
import React from 'react';
import { configure, mount, ReactWrapper } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import ChartElement, {
  ChartElement as PureChartElement
} from '../chart_element/chartelement';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import {
  MockModel,
  recorderData,
  computedResult,
  driverData
} from '../../../utils/tests/utils';
import { StateInterface } from '../../redux/types';
configure({ adapter: new Adapter() });

const mockStore = configureMockStore([thunk]);

describe('Test <ChartElement/>', () => {
  let wrapper: ReactWrapper;
  let btnHandle;
  let divRef;

  beforeEach(() => {
    const newState: StateInterface = {
      ...initialState,
      dashboardState: {
        ...initialState.dashboardState,
        computedResult,
        recorderData,
        driverData
      }
    };
    const store = mockStore(newState);
    btnHandle = jest.fn(() => {});
    divRef = React.createRef<any>();
    wrapper = mount(
      <Provider store={store}>
        <ChartElement
          ref={divRef}
          model={new MockModel() as any}
          send_msg={jest.fn()}
          id={'id'}
          initialState={null}
        />
      </Provider>
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('Should render correctly component', () => {
    const component = wrapper.find(ChartElement);
    expect(component).toHaveLength(1);
  });

  it('Should match snapshot', () => {
    const component = wrapper.find(ChartElement);
    expect(component.html()).toMatchSnapshot();
  });

  it('Should create state correctly ', () => {
    const component = wrapper.find(PureChartElement);
    expect(component.state('plotLayoutConfig')).toEqual({
      title: '',
      xAxis: '',
      xAxisMax: '',
      xAxisMin: '',
      yAxis: '',
      yAxisMax: '',
      yAxisMin: '',
      zAxis: '',
      zAxisMax: '',
      zAxisMin: ''
    });
    expect(component.state('systemNameX')).toEqual('model.source');
    expect(component.state('xData')).toEqual('inwards.I');
    expect(component.state('listSystem')).toEqual([
      'model.source',
      'model.ground',
      'model.circuit',
      'model.circuit.n1',
      'model.circuit.n2',
      'model.circuit.R1',
      'model.circuit.R2',
      'model.circuit.R3'
    ]);
    expect(component.state('dictSystemVariable')).toEqual({
      'model.circuit': ['inwards.n1_V', 'inwards.n2_V', 'I_in.I', 'Vg.V'],
      'model.circuit.R1': [
        'inwards.R',
        'V_in.V',
        'V_out.V',
        'outwards.deltaV',
        'I.I'
      ],
      'model.circuit.R2': [
        'inwards.R',
        'V_in.V',
        'V_out.V',
        'outwards.deltaV',
        'I.I'
      ],
      'model.circuit.R3': [
        'inwards.R',
        'V_in.V',
        'V_out.V',
        'outwards.deltaV',
        'I.I'
      ],
      'model.circuit.n1': [
        'inwards.n_in',
        'inwards.n_out',
        'inwards.V',
        'I_in0.I',
        'I_out0.I',
        'I_out1.I',
        'outwards.sum_I_in',
        'outwards.sum_I_out'
      ],
      'model.circuit.n2': [
        'inwards.n_in',
        'inwards.n_out',
        'inwards.V',
        'I_in0.I',
        'I_out0.I',
        'outwards.sum_I_in',
        'outwards.sum_I_out'
      ],
      'model.ground': ['inwards.V', 'V_out.V'],
      'model.source': ['inwards.I', 'I_out.I']
    });
    expect(component.state('dictSystemVariableLength')).toEqual({
      'model.circuit': {
        'I_in.I': ['All'],
        'Vg.V': ['All'],
        'inwards.n1_V': ['All'],
        'inwards.n2_V': ['All']
      },
      'model.circuit.R1': {
        'I.I': ['All'],
        'V_in.V': ['All'],
        'V_out.V': ['All'],
        'inwards.R': ['All'],
        'outwards.deltaV': ['All']
      },
      'model.circuit.R2': {
        'I.I': ['All'],
        'V_in.V': ['All'],
        'V_out.V': ['All'],
        'inwards.R': ['All'],
        'outwards.deltaV': ['All']
      },
      'model.circuit.R3': {
        'I.I': ['All'],
        'V_in.V': ['All'],
        'V_out.V': ['All'],
        'inwards.R': ['All'],
        'outwards.deltaV': ['All']
      },
      'model.circuit.n1': {
        'I_in0.I': ['All'],
        'I_out0.I': ['All'],
        'I_out1.I': ['All'],
        'inwards.V': ['All'],
        'inwards.n_in': ['All'],
        'inwards.n_out': ['All'],
        'outwards.sum_I_in': ['All'],
        'outwards.sum_I_out': ['All']
      },
      'model.circuit.n2': {
        'I_in0.I': ['All'],
        'I_out0.I': ['All'],
        'inwards.V': ['All'],
        'inwards.n_in': ['All'],
        'inwards.n_out': ['All'],
        'outwards.sum_I_in': ['All'],
        'outwards.sum_I_out': ['All']
      },
      'model.ground': { 'V_out.V': ['All'], 'inwards.V': ['All'] },
      'model.source': { 'I_out.I': ['All'], 'inwards.I': ['All'] }
    });
    expect(component.state('listRecorder')).toEqual(['model.design']);
    expect(component.state('dictRecorderVariable')).toEqual({
      'model.design': [
        'circuit.R1.R',
        'circuit.R2.R',
        'circuit.n1.V',
        'circuit.n2.V',
        'ground.V',
        'source.I'
      ]
    });
    expect(component.state('dictRecorderVariableLength')).toEqual({
      'model.design': {
        'circuit.R1.R': ['All'],
        'circuit.R2.R': ['All'],
        'circuit.n1.V': ['All'],
        'circuit.n2.V': ['All'],
        'ground.V': ['All'],
        'source.I': ['All']
      }
    });
    expect(component.state('dictRecorderRef')).toEqual({
      'model.design': ['All', 'pt1', 'pt2']
    });
    expect(component.state('listDriverWithTrace')).toEqual(['model.design']);
    expect(component.state('dictDriverVariable')).toEqual({
      'model.design': ['Residue']
    });
  });
  it('Should open setting dialog', () => {
    const component = wrapper.find(PureChartElement);
    const instance = component.instance() as PureChartElement;
    instance.openSetting();
    expect(component.state('openSetting')).toEqual(true);
    expect(component.state('dialogType')).toEqual('new');
  });
  it('Should close setting dialog', () => {
    const component = wrapper.find(PureChartElement);
    const instance = component.instance() as PureChartElement;
    instance.closeDialog();
    expect(component.state('openSetting')).toEqual(false);
  });
  it('Should open layout setting dialog', () => {
    const component = wrapper.find(PureChartElement);
    const instance = component.instance() as PureChartElement;
    instance.openLayoutSetting();
    expect(component.state('layoutSettingDialog')).toEqual(true);
  });
  it('Should update store after changing data source to Variables', () => {
    const component = wrapper.find(PureChartElement);
    const instance = component.instance() as PureChartElement;
    instance.handleDatasourceChange({ target: { value: 'Variables' } });
    expect(component.state('dataSource')).toEqual('Variables');
    expect(component.state('listSelector')).toEqual([
      'model.source',
      'model.ground',
      'model.circuit',
      'model.circuit.n1',
      'model.circuit.n2',
      'model.circuit.R1',
      'model.circuit.R2',
      'model.circuit.R3'
    ]);
    expect(component.state('dictSelector')).toEqual({
      'model.circuit': ['inwards.n1_V', 'inwards.n2_V', 'I_in.I', 'Vg.V'],
      'model.circuit.R1': [
        'inwards.R',
        'V_in.V',
        'V_out.V',
        'outwards.deltaV',
        'I.I'
      ],
      'model.circuit.R2': [
        'inwards.R',
        'V_in.V',
        'V_out.V',
        'outwards.deltaV',
        'I.I'
      ],
      'model.circuit.R3': [
        'inwards.R',
        'V_in.V',
        'V_out.V',
        'outwards.deltaV',
        'I.I'
      ],
      'model.circuit.n1': [
        'inwards.n_in',
        'inwards.n_out',
        'inwards.V',
        'I_in0.I',
        'I_out0.I',
        'I_out1.I',
        'outwards.sum_I_in',
        'outwards.sum_I_out'
      ],
      'model.circuit.n2': [
        'inwards.n_in',
        'inwards.n_out',
        'inwards.V',
        'I_in0.I',
        'I_out0.I',
        'outwards.sum_I_in',
        'outwards.sum_I_out'
      ],
      'model.ground': ['inwards.V', 'V_out.V'],
      'model.source': ['inwards.I', 'I_out.I']
    });
    expect(component.state('systemNameX')).toEqual('model.source');
    expect(component.state('xData')).toEqual('inwards.I');
    expect(component.state('xDataIndex')).toEqual('All');
  });
  it('Should update store after changing data source to Recorders', () => {
    const component = wrapper.find(PureChartElement);
    const instance = component.instance() as PureChartElement;
    instance.handleDatasourceChange({ target: { value: 'Recorders' } });
    expect(component.state('dataSource')).toEqual('Recorders');
    expect(component.state('listSelector')).toEqual(['model.design']);
    expect(component.state('dictSelector')).toEqual({
      'model.design': [
        'circuit.R1.R',
        'circuit.R2.R',
        'circuit.n1.V',
        'circuit.n2.V',
        'ground.V',
        'source.I'
      ]
    });
    expect(component.state('systemNameX')).toEqual('model.design');
    expect(component.state('xData')).toEqual('circuit.R1.R');
    expect(component.state('xDataIndex')).toEqual('All');
  });
});
