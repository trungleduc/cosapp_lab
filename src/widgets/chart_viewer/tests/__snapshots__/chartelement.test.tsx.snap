// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <ChartElement/> Should match snapshot 1`] = `"<div class=\\"cosapp-widget-box\\"><div style=\\"width: 100%;\\"></div><div class=\\"ForwardRef-bottomBarDiv-24\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\" style=\\"color: rgb(50, 50, 50);\\"><span class=\\"MuiButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\" style=\\"color: rgb(50, 50, 50);\\"><span class=\\"MuiButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\" style=\\"color: rgb(50, 50, 50);\\"><span class=\\"MuiButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M2 20h20v-4H2v4zm2-3h2v2H4v-2zM2 4v4h20V4H2zm4 3H4V5h2v2zm-4 7h20v-4H2v4zm2-3h2v2H4v-2z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\" style=\\"color: rgb(50, 50, 50);\\"><span class=\\"MuiButton-label\\"> <svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div>"`;
